server/src/main/java/com/defold/extender/services/DefoldSdkService.java

```diff
@@ -109,26 +96,45 @@ public boolean isLocalSdk(final String sdkVersion) {
         return sdkVersion == null || LOCAL_VERSION.equals(sdkVersion) || System.getenv("DYNAMO_HOME") != null;
     }
 
-    public File getSdk(String hash) throws IOException, URISyntaxException, ExtenderException {
-        return isLocalSdk(hash) ? getLocalSdk() : getRemoteSdk(hash);
+    public DefoldSdk getSdk(String hash) throws ExtenderException {
+        if (isLocalSdk(hash)){
+            return getLocalSdk();
+        }
+        // use ConcurrentHashMap with CompletableFuture here to avoid situation when
+        // several builds needs the same defoldsdk which doesn't exist locally. So one build job starts downloading,
+        // all other jobs wait for download complete and all of then continue running.
+        CompletableFuture<DefoldSdk> operation = operationCache.computeIfAbsent(hash, this::getRemoteSdk);
+        try (DefoldSdk sdk = operation.get()){
+            if (sdk == null) {
+                throw new ExtenderException(String.format("The given sdk does not exist: %s", hash));
+            }
+            if (!sdk.isValid()) {
+                throw new ExtenderException(String.format("The given sdk still does not exist: %s", hash));
```

Comment:
I'd argue that it exists, but is not valid?

Reply:
Seems you right and it's an obsolete condition. I've moved it from old code where we try synchronize different thread via `Thread.sleep`. So I'll remove it. Thanks!